app/main.py

```diff
@@ -1,34 +1,210 @@
+                break
+        else:
+            self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple]) -> None:
+        self.ships = ships
+        self.field = dict()
+        self.create_ships()
+        if not self._validate_field():
+            print("You should re_create your ships.")
+            self.field = dict()
+
+    def fire(self, location: tuple) -> str:
+        if location in self.field:
+            ship = self.field.get(location)
+            ship.fire(row=location[0], column=location[1])
+            if ship.is_drowned:
+                return "Sunk!"
+            return "Hit!"
+        return "Miss!"
+
+    def create_ships(self) -> None:
+        ships = []
+        for location in self.ships:
+            ships.append(Ship(start=location[0], end=location[1]))
+
+        for ship in ships:
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    def print_field(self) -> None:
+        whole_field = dict()
+        for i in range(10):
+            for index in range(10):
+                symbol = "~"
+                if (i, index) in self.field:
+                    symbol = "□"
+                whole_field[(i, index)] = symbol
+
+        for location in whole_field:
+            ship = self.field.get(location)
+            if ship:
+                symbol = "*"
+                if ship.is_drowned:
+                    symbol = "x"
+                for deck in ship.decks:
+                    if not deck.is_alive:
+                        whole_field[(deck.row, deck.column)] = symbol
+
+        for index, location in enumerate(whole_field):
+            if index % 10 == 0:
+                print("")
+            print(f"{whole_field.get(location)}      ", end="")
+
+    def _validate_field(self) -> bool:
+        single_deck_ship = 0
+        double_deck_ship = 0
+        three_deck_ship = 0
+        four_deck_ship = 0
+
+        ships = set([ship for ship in self.field.values()])
+        for ship in ships:
+            ship_length = len(ship.decks)
+            if ship_length == 1:
+                single_deck_ship += 1
+                continue
+            if ship_length == 2:
+                double_deck_ship += 1
+                continue
+            if ship_length == 3:
+                three_deck_ship += 1
+                continue
+            if ship_length == 4:
+                four_deck_ship += 1
```

Comment:
Consider using `dict` here to store ship length as the key and the number of ships with particular lengths as values. Increment `dict` values instead of 4 variables, it will avoid the use of 4 `if` conditions.